custom_components/opnsense/device_tracker.py

```
@@ -36,12 +36,183 @@
+        if not isinstance(arp_entry, MutableMapping):
+            continue
+        mac_address = arp_entry.get("mac")
+        if not isinstance(mac_address, str) or not mac_address or mac_address in mac_addresses:
+            continue
+        mac_addresses.append(mac_address)
+        devices.append(_device_data_from_arp_entry(mac_address, arp_entry))
+
+    return devices, mac_addresses
+
+
+def _non_empty_string(value: Any) -> str | None:
```

Ruff (0.15.18)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `value`

(ANN401)